mono/metadata/class-internals.h

```diff
@@ -1067,6 +1067,22 @@ mono_register_jit_icall (gconstpointer func, const char *name, MonoMethodSignatu
 MonoJitICallInfo *
 mono_register_jit_icall_full (gconstpointer func, const char *name, MonoMethodSignature *sig, gboolean no_wrapper, const char *c_symbol);
 
+#ifdef __cplusplus
+template <typename T>
+inline MonoJitICallInfo *
+mono_register_jit_icall (T func, const char *name, MonoMethodSignature *sig, gboolean is_save)
```

Comment:
One day, maybe, we can make this more typesafe than the C version:

```c++
template <typename T>
inline typename std::enable_if<std::is_function <T>::value, MonoJitICallInfo*>::type
mono_register_jit_icall (T func, const char *name, MonoMethodSignature *sig, gboolean is_save)
{
   return mono_register_jit_icall ((gconstpointer)func, name, sig, is_save);
}
```

Reply:
Agreed but not today.
Does it end up being templates all the way down, or is this the correct C++ approach?

Reply:
in the libraries that i've seen, sooner or later the templates cease and you end up at some untyped storage layer where everything's a `char*` (and a destructor).

So for something that's essentially a big lookup table of heterogeneous functions this is probably close to the right place to cut down to a `gconstpointer`.